### PR TITLE
[Merged by Bors] - chore(Ideal/Operations): remove extraneous nonempty conditions

### PR DESCRIPTION
**Pull request by alreadydone**
Also generalizes `Ideal.subset_union` to `Subgroup`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 1]
```diff
@@ -191,6 +191,14 @@ theorem mul_mem_cancel_left {x y : G} (h : x ∈ H) : x * y ∈ H ↔ y ∈ H :=
...
 
+@[to_additive]
+theorem SubgroupClass.subset_union {H K L : S} : (H : Set G) ⊆ K ∪ L ↔ H ≤ K ∨ H ≤ L := by
+  refine ⟨fun h ↦ ?_, fun h ↦ h.elim (fun h _ hH ↦ .inl (h hH)) fun h _ hH ↦ .inr (h hH)⟩
+  rw [or_iff_not_imp_left, SetLike.not_le_iff_exists]
+  exact fun ⟨x, xH, xK⟩ y yH ↦ (h <| mul_mem xH yH).elim
+    ((h yH).resolve_left fun yK ↦ xK <| (mul_mem_cancel_right yK).mp ·)
+    (mul_mem_cancel_left <| (h xH).resolve_left xK).mp
+
 namespace SubgroupClass
```
Comment (alreadydone) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:202]:
```suggestion
namespace SubgroupClass

@[to_additive]
theorem subset_union {H K L : S} : (H : Set G) ⊆ K ∪ L ↔ H ≤ K ∨ H ≤ L := by
  refine ⟨fun h ↦ ?_, fun h x xH ↦ h.imp (· xH) (· xH)⟩
  rw [or_iff_not_imp_left, SetLike.not_le_iff_exists]
  exact fun ⟨x, xH, xK⟩ y yH ↦ (h <| mul_mem xH yH).elim
    ((h yH).resolve_left fun yK ↦ xK <| (mul_mem_cancel_right yK).mp ·)
    (mul_mem_cancel_left <| (h xH).resolve_left xK).mp
    
```

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 2]
```diff
@@ -193,6 +193,15 @@ theorem mul_mem_cancel_left {x y : G} (h : x ∈ H) : x * y ∈ H ↔ y ∈ H :=
...
 
+@[to_additive]
+theorem subset_union {H K L : S} : (H : Set G) ⊆ K ∪ L ↔ H ≤ K ∨ H ≤ L := by
+  refine ⟨fun h ↦ ?_, fun h x xH ↦ h.imp (· xH) (· xH)⟩
+  rw [or_iff_not_imp_left, SetLike.not_le_iff_exists]
+  exact fun ⟨x, xH, xK⟩ y yH ↦ (h <| mul_mem xH yH).elim
+    ((h yH).resolve_left fun yK ↦ xK <| (mul_mem_cancel_right yK).mp ·)
+    (mul_mem_cancel_left <| (h xH).resolve_left xK).mp
+
```
Comment (alreadydone) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:203]:
```suggestion
```

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 3]
```diff
@@ -177,22 +177,28 @@ theorem exists_inv_mem_iff_exists_mem {P : G → Prop} :
...
   ⟨fun hab => by simpa using mul_mem (inv_mem h) hab, mul_mem h⟩
 #align mul_mem_cancel_left mul_mem_cancel_left
 #align add_mem_cancel_left add_mem_cancel_left
 
 namespace SubgroupClass
 
+-- Here we assume H, K, and L are subgroups, but in fact one of them could
+-- be allowed to be a submonoid only.
+@[to_additive]
+theorem subset_union {H K L : S} : (H : Set G) ⊆ K ∪ L ↔ H ≤ K ∨ H ≤ L := by
```
Comment (erdOne) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:195]: I wonder if a finite union version could be useful?

Comment (erdOne) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:195]: Nah. Maybe not.

**Comment by erdOne** [comment]
maintainer merge

File: Mathlib/GroupTheory/Subgroup/Basic.lean (hunk 3 above)
Comment (alreadydone) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:195]: It's not true when you have a union of three subgroups and the simplest example is the three order 2 subgroups in a Klein four group whose union is the whole group.

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 4]
```diff
@@ -177,22 +177,28 @@ theorem exists_inv_mem_iff_exists_mem {P : G → Prop} :
...
-  -- Porting note: whut? why do we need this?
-  have : SubmonoidClass S G := SubgroupClass.toSubmonoidClass
   ⟨fun hab => by simpa using mul_mem (inv_mem h) hab, mul_mem h⟩
 #align mul_mem_cancel_left mul_mem_cancel_left
 #align add_mem_cancel_left add_mem_cancel_left
 
 namespace SubgroupClass
 
+-- Here we assume H, K, and L are subgroups, but in fact one of them could
+-- be allowed to be a submonoid only.
```
Comment (alreadydone) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:193]:
```suggestion
-- Here we assume H, K, and L are subgroups, but in fact any one of them
-- could be allowed to be a subsemigroup.
-- Counterexample where K and L are submonoids: H = ℤ, K = ℕ, L = -ℕ
-- Counterexample where H and K are submonoids: H = {n | n = 0 ∨ 3 ≤ n}, K = 3ℕ + 4ℕ, L = 5ℤ 
```

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**